### PR TITLE
Trace for NNUE eval.

### PR DESCRIPTION
**Pull request by Sopel97**
Some WIP, want to know if it's desired to have.

This patch adds some more output to the `eval` command. It adds a board display with estimated piece values (remove piece, eval, put piece), and split psqt+position eval for each bucket. Example:

```
Stockfish 170621 by the Stockfish developers (see AUTHORS file)
eval
info string NNUE evaluation using nn-33c9d39e5eb6.nnue enabled

     Term    |    White    |    Black    |    Total
             |   MG    EG  |   MG    EG  |   MG    EG
 ------------+-------------+-------------+------------
    Material |  ----  ---- |  ----  ---- |  0.00  0.00
   Imbalance |  ----  ---- |  ----  ---- |  0.00  0.00
       Pawns |  0.38 -0.08 |  0.38 -0.08 |  0.00  0.00
     Knights | -0.02 -0.19 | -0.02 -0.19 |  0.00  0.00
     Bishops |  0.01 -0.41 |  0.01 -0.41 |  0.00  0.00
       Rooks | -0.26 -0.06 | -0.26 -0.06 |  0.00  0.00
      Queens |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Mobility | -0.88 -1.15 | -0.88 -1.15 |  0.00  0.00
 King safety |  0.91 -0.10 |  0.91 -0.10 |  0.00  0.00
     Threats |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
      Passed |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Space |  0.40  0.00 |  0.40  0.00 |  0.00  0.00
    Winnable |  ----  ---- |  ----  ---- |  0.00  0.00
 ------------+-------------+-------------+------------
       Total |  ----  ---- |  ----  ---- |  0.00  0.00

+-------+-------+-------+-------+-------+-------+-------+-------+
|   r   |   n   |   b   |   q   |   k   |   b   |   n   |   r   |
| -7.83 | -7.63 | -7.68 | -13.6 |       | -8.06 | -7.71 | -8.25 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   p   |   p   |   p   |   p   |   p   |   p   |   p   |   p   |
| -1.04 | -1.35 | -1.50 | -1.48 | -1.57 | -1.91 | -1.79 | -1.00 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   P   |   P   |   P   |   P   |   P   |   P   |   P   |   P   |
| +0.89 | +1.22 | +1.28 | +1.26 | +1.37 | +1.73 | +1.64 | +0.71 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   R   |   N   |   B   |   Q   |   K   |   B   |   N   |   R   |
| +7.44 | +6.86 | +7.09 | +11.7 |       | +7.16 | +6.55 | +7.69 |
+-------+-------+-------+-------+-------+-------+-------+-------+

     | Bckt  | PSQT  |  LS   | Total |
     |     0 |  0.00 | -0.89 | -0.89 |
     |     1 |  0.00 | +0.36 | +0.36 |
     |     2 |  0.00 | +0.20 | +0.20 |
     |     3 |  0.00 | +0.21 | +0.21 |
     |     4 |  0.00 | +0.17 | +0.17 |
     |     5 |  0.00 | +0.11 | +0.11 |
     |     6 |  0.00 | +0.17 | +0.17 |
 --> |     7 |  0.00 | +0.16 | +0.16 |


Classical evaluation: 0.00 (white side)

NNUE evaluation:      0.17 (white side)

Final evaluation:     0.31 (white side)

position fen r1bqkbnr/ppp2ppp/2np4/4p3/4P3/3B1N2/PPPP1PPP/RNBQ1RK1 b kq - 3 4
eval
info string NNUE evaluation using nn-33c9d39e5eb6.nnue enabled

     Term    |    White    |    Black    |    Total
             |   MG    EG  |   MG    EG  |   MG    EG
 ------------+-------------+-------------+------------
    Material |  ----  ---- |  ----  ---- |  0.80  0.00
   Imbalance |  ----  ---- |  ----  ---- |  0.00  0.00
       Pawns |  0.34 -0.07 |  0.59 -0.02 | -0.25 -0.05
     Knights | -0.18 -0.29 | -0.07 -0.16 | -0.12 -0.13
     Bishops | -0.46 -1.39 | -0.14 -1.02 | -0.32 -0.37
       Rooks |  0.03 -0.02 | -0.19 -0.06 |  0.22  0.04
      Queens |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Mobility | -0.11 -0.14 |  0.08  0.21 | -0.19 -0.36
 King safety |  0.61 -0.10 |  0.68 -0.10 | -0.08  0.00
     Threats |  0.13  0.13 |  0.17  0.17 | -0.03 -0.03
      Passed |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Space |  0.54  0.00 |  0.61  0.00 | -0.07  0.00
    Winnable |  ----  ---- |  ----  ---- |  0.00 -0.28
 ------------+-------------+-------------+------------
       Total |  ----  ---- |  ----  ---- | -0.03 -1.17

+-------+-------+-------+-------+-------+-------+-------+-------+
|   r   |       |   b   |   q   |   k   |   b   |   n   |   r   |
| +6.80 |       | +7.01 | +12.4 |       | +6.66 | +6.14 | +6.97 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   p   |   p   |   p   |       |       |   p   |   p   |   p   |
| +0.82 | +1.33 | +1.29 |       |       | +1.16 | +0.94 | +0.37 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |   n   |   p   |       |       |       |       |
|       |       | +6.92 | +1.08 |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |   p   |       |       |       |
|       |       |       |       | +1.89 |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |   P   |       |       |       |
|       |       |       |       | -1.78 |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |   B   |       |   N   |       |       |
|       |       |       | -8.12 |       | -7.93 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   P   |   P   |   P   |   P   |       |   P   |   P   |   P   |
| -0.81 | -1.22 | -1.32 | -1.01 |       | -1.34 | -2.96 | -1.54 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   R   |   N   |   B   |   Q   |       |   R   |   K   |       |
| -7.93 | -7.07 | -7.69 | -13.4 |       | -8.18 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+

     | Bckt  | PSQT  |  LS   | Total |
     |     0 | +0.23 | -0.08 | +0.14 |
     |     1 | -0.08 | +0.35 | +0.26 |
     |     2 | -0.33 | +0.32 | -0.00 |
     |     3 | -0.39 | +0.27 | -0.12 |
     |     4 | -0.28 | +0.45 | +0.16 |
     |     5 | -0.14 | +0.32 | +0.17 |
     |     6 | -0.12 | +0.18 | +0.05 |
 --> |     7 | -0.13 | +0.14 | +0.00 |


Classical evaluation: 0.00 (white side)

NNUE evaluation:      -0.00 (white side)

Final evaluation:     -0.03 (white side)

position fen 8/3k4/5P2/5K2/8/8/8/8 w - - 0 1
eval
info string NNUE evaluation using nn-33c9d39e5eb6.nnue enabled

     Term    |    White    |    Black    |    Total
             |   MG    EG  |   MG    EG  |   MG    EG
 ------------+-------------+-------------+------------
    Material |  ----  ---- |  ----  ---- |  0.00  0.00
   Imbalance |  ----  ---- |  ----  ---- |  0.00  0.00
       Pawns |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Knights |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Bishops |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Rooks |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
      Queens |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Mobility |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
 King safety |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Threats |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
      Passed |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Space |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Winnable |  ----  ---- |  ----  ---- |  0.00  0.00
 ------------+-------------+-------------+------------
       Total |  ----  ---- |  ----  ---- |  0.00  0.00

+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |   k   |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   P   |       |       |
|       |       |       |       |       | +11.6 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   K   |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+

     | Bckt  | PSQT  |  LS   | Total |
 --> |     0 | +0.93 | +10.7 | +11.6 |
     |     1 | +0.19 | +7.96 | +8.15 |
     |     2 | +0.74 | +4.76 | +5.51 |
     |     3 | +1.07 | +2.92 | +4.00 |
     |     4 | +1.11 | +0.78 | +1.90 |
     |     5 | +0.31 | +2.25 | +2.57 |
     |     6 | +0.16 | +3.75 | +3.91 |
     |     7 | +0.38 | +3.11 | +3.50 |


Classical evaluation: 49.10 (white side)

NNUE evaluation:      11.70 (white side)

Final evaluation:     11.12 (white side)

position fen 8/5k2/5P2/5K2/8/8/8/8 w - - 0 1
eval
info string NNUE evaluation using nn-33c9d39e5eb6.nnue enabled

     Term    |    White    |    Black    |    Total
             |   MG    EG  |   MG    EG  |   MG    EG
 ------------+-------------+-------------+------------
    Material |  ----  ---- |  ----  ---- |  0.00  0.00
   Imbalance |  ----  ---- |  ----  ---- |  0.00  0.00
       Pawns |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Knights |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Bishops |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Rooks |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
      Queens |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Mobility |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
 King safety |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
     Threats |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
      Passed |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
       Space |  0.00  0.00 |  0.00  0.00 |  0.00  0.00
    Winnable |  ----  ---- |  ----  ---- |  0.00  0.00
 ------------+-------------+-------------+------------
       Total |  ----  ---- |  ----  ---- |  0.00  0.00

+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   k   |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   P   |       |       |
|       |       |       |       |       | -0.14 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   K   |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+

     | Bckt  | PSQT  |  LS   | Total |
 --> |     0 | +0.92 | -0.93 | -0.00 |
     |     1 | +0.13 | +1.41 | +1.54 |
     |     2 | +0.60 | +0.58 | +1.18 |
     |     3 | +0.84 | -0.71 | +0.13 |
     |     4 | +0.84 | -0.62 | +0.21 |
     |     5 | +0.24 | +0.16 | +0.40 |
     |     6 | +0.09 | +2.50 | +2.60 |
     |     7 | +0.33 | +1.37 | +1.70 |


Classical evaluation: 0.00 (white side)

NNUE evaluation:      -0.01 (white side)

Final evaluation:     -0.10 (white side)
```

### CODE REVIEW

File: src/nnue/evaluate_nnue.cpp [hunk 1]
```diff
@@ -175,6 +177,148 @@ namespace Stockfish::Eval::NNUE {
...
+    return t;
+  }
+
+  static const std::string PieceToChar(" PNBRQK  pnbrqk");
+
+  static void format_cp_5(Value v, char* buffer) {
+    buffer[0] = (v < 0 ? '-' : v > 0 ? '+' : ' ');
+    int cp = (int)(abs((float)v / PawnValueEg * 100.0f + 0.5f));
+    if (cp >= 10000) {
+      buffer[1] = '0' + cp / 10000; cp %= 10000;
```
[the remark this answers is not in the document]
Comment (Sopel97) [reply src/nnue/evaluate_nnue.cpp:233]: yep mistake

**Comment by snicolet** [comment]
I like the idea. Some remarks:

- wondering if you could use some `<iomanip>` and `setprecision` magic to write the values
- this feature, if merged, is going to be used widely, by people not in the development team. I suggest to use very explicit headers for the columns, to lower the number of questions we'll receive:

```
Bckt --> Bucket
PSQT --> Material
LS --> Positional (or whatever is adequate, important point is to use a word that fans can understand)
Total --> Total
```

Also, the table could be left aligned, the arrow could be on the right of the table and more explicit:

```
 | Bckt  | PSQT  |  LS   | Total |
 |     0 | +0.92 | -0.93 | -0.00 |
 |     1 | +0.13 | +1.41 | +1.54 |    
 |     2 | +0.60 | +0.58 | +1.18 |
 |     3 | +0.84 | -0.71 | +0.13 |     <--- this bucket is used
 |     4 | +0.84 | -0.62 | +0.21 |
 |     5 | +0.24 | +0.16 | +0.40 |
 |     6 | +0.09 | +2.50 | +2.60 |
 |     7 | +0.33 | +1.37 | +1.70 |
```

**Comment by vondele** [comment]
probably calling PSQT material is a bit misleading, there is a lot of positional know-how in that part. It is like mean-field positional understanding (i.e. only positions of single pieces matter) whereas our LS (Layer Stack) provides more the 'correlation' part (multiple pieces matter9 of it. (at least that's how I sometimes think about it).

**Comment by snicolet** [comment]
I agree that the PSQT certainly have some positional understanding of pieces locations (especially with respect to the kings positions). After all, it is "simulating" the classic feature "Positional square tables", just on steroids.

on the other hand, I have the intuition that *on average* positions with a one pawn advantage will have a +1 PSQT, two pawnw advantage will give +2 PSQT, and one rook advantage will give +5 PSQT.

**Comment by Sopel97** [comment]
I went for a compromise and included both wordings. Also changed the formatting of the table and the values in it slightly.

> wondering if you could use some <iomanip> and setprecision magic to write the values

Maybe, but personally I hate that stuff, never know if it stays and needs to be reset or now. If someone makes it work in the future I'm fine with that. Also now I'm not sure it's applicable to the table as there dots are vertically aligned.

```
+------------+------------+------------+------------+
|   Bucket   |  Material  | Positional |   Total    |
|            |   (PSQT)   |  (Layers)  |            |
+------------+------------+------------+------------+
|  0         |     0.00   |  -  0.88   |  -  0.88   |
|  1         |     0.00   |  +  0.36   |  +  0.36   |
|  2         |     0.00   |  +  0.21   |  +  0.21   |
|  3         |     0.00   |  +  0.22   |  +  0.22   |
|  4         |     0.00   |  +  0.17   |  +  0.17   |
|  5         |     0.00   |  +  0.12   |  +  0.12   |
|  6         |     0.00   |  +  0.17   |  +  0.17   |
|  7         |     0.00   |  +  0.17   |  +  0.17   | <-- this bucket is used
+------------+------------+------------+------------+
```

**Comment by vondele** [comment]
> But pawns on the sixth rank will get +2 PSQT, probably that what you mean by "mean-field positional understanding"?

yes, or similar things like rooks on the 7th rank, knights on the rim, bishops controlling the centre, etc.. all just properties of single pieces.

**Comment by snicolet** [comment]
I think we could use the opportunity of this PR to remove the following line:

```
Final evaluation:     -0.10 (white side)
```

It is just disturbing and a source of endless questioning. The two evals (classical and NNUE) would suffice to cover the curiosity of the fans

**Comment by Sopel97** [comment]
I think it's important to see the real evaluation the engine sees, in particular how it's adjusted.

**Comment by snicolet** [comment]
I have no idea what "Final evaluation" is ? 
why is it different from NNUE eval since I have used the "Use NNUE option" in the UCI commands?

**Comment by vondele** [comment]
'adjusted_NNUE()'-scaling and similar. It is the real evaluation used in search.

**Comment by snicolet** [comment]
I know -- I was just wondering who in the team should give his telephone number to answer these kind of questions to the fans

**Comment by Sopel97** [comment]
If something is unclear then it's an issue with `eval` documentation and not really what it outputs

**Comment by vondele** [comment]
so should we have little titles for the 'tables' like

`contributions to the classical eval`
`NNUE derived piece values`
`NNUE network contributions`

**Comment by snicolet** [comment]
I am working on merging this patch

**Comment by vdbergh** [comment]
If this really needs to be understandable to people not familiar with the source code (some posts above indicate that this is the goal) then some things are quite baffling. E.g. consider 8/3k4/5P2/5K2/8/8/8/8 w - - 0 1 (the case of won KPK endgame).

(1) All entries in the classical evaluation table are zero. Nonetheless the classical evaluation is not zero (I guess this has to do with the fact that a special function is invoked, but this is not indicated). 
(2) Both the diagram and the bucket give the value 11.6 (duplicate information?) but the NNUE eval is 11.70..
(3) There is no indication how the final evaluation is computed. I guess some scale factors are involved, but which ones?

**Comment by snicolet** [comment]
Example of output for my current version:

```
./stockfish    
position fen 3Qb1k1/1r2ppb1/pN1n2q1/Pp1Pp1Pr/4P2p/4BP2/4B1R1/1R5K b - - 11 40
eval
```

```
info string NNUE evaluation using nn-50144f835024.nnue enabled

 Contributing terms for the classical eval:
+------------+-------------+-------------+-------------+
|    Term    |    White    |    Black    |    Total    |
|            |   MG    EG  |   MG    EG  |   MG    EG  |
+------------+-------------+-------------+-------------+
|   Material |  ----  ---- |  ----  ---- | -0.73 -1.55 |
|  Imbalance |  ----  ---- |  ----  ---- | -0.21 -0.17 |
|      Pawns |  0.35 -0.00 |  0.19 -0.26 |  0.16  0.25 |
|    Knights |  0.04 -0.08 |  0.12 -0.01 | -0.08 -0.07 |
|    Bishops | -0.34 -0.87 | -0.17 -0.61 | -0.17 -0.26 |
|      Rooks |  0.12  0.00 |  0.08  0.00 |  0.04  0.00 |
|     Queens |  0.00  0.00 | -0.27 -0.07 |  0.27  0.07 |
|   Mobility |  0.84  1.76 |  0.01  0.66 |  0.83  1.10 |
|King safety | -0.99 -0.17 | -0.72 -0.10 | -0.27 -0.07 |
|    Threats |  0.27  0.27 |  0.73  0.86 | -0.46 -0.59 |
|     Passed |  0.00  0.00 |  0.79  0.82 | -0.79 -0.82 |
|      Space |  0.61  0.00 |  0.24  0.00 |  0.37  0.00 |
|   Winnable |  ----  ---- |  ----  ---- |  0.00 -0.03 |
+------------+-------------+-------------+-------------+
|      Total |  ----  ---- |  ----  ---- | -1.03 -2.14 |
+------------+-------------+-------------+-------------+

 Pure NNUE derived piece values:
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |   Q   |   b   |       |   k   |       |
|       |       |       | +12.4 | -1.61 |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |   r   |       |       |   p   |   p   |   b   |       |
|       | -3.88 |       |       | -0.84 | -1.18 | -3.31 |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   p   |   N   |       |   n   |       |       |   q   |       |
| -1.80 | +3.72 |       | -4.82 |       |       | -5.03 |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|   P   |   p   |       |   P   |   p   |       |   P   |   r   |
| +1.16 | -0.90 |       | +0.56 | +0.12 |       | +0.51 | -4.02 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |   P   |       |       |   p   |
|       |       |       |       | +2.33 |       |       | +1.17 |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |   B   |   P   |       |       |
|       |       |       |       | +4.79 | +1.55 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |   B   |       |   R   |       |
|       |       |       |       | +4.54 |       | +6.03 |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |   R   |       |       |       |       |       |   K   |
|       | +4.81 |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+

 Pure NNUE network contributions:
+------------+------------+------------+------------+
|   Bucket   |  Material  | Positional |   Total    |
|            |   (PSQT)   |  (Layers)  |            |
+------------+------------+------------+------------+
|  0         |  +  0.33   |  -  1.46   |  -  1.13   |
|  1         |  +  0.25   |  -  0.67   |  -  0.42   |
|  2         |  +  0.47   |  -  1.72   |  -  1.25   |
|  3         |  +  0.56   |  -  1.80   |  -  1.24   |
|  4         |  +  0.48   |  -  1.77   |  -  1.29   |
|  5         |  +  0.40   |  -  2.00   |  -  1.60   |
|  6         |  +  0.58   |  -  2.11   |  -  1.53   | <-- this bucket is used
|  7         |  +  3.38   |  -  2.00   |  +  1.38   |
+------------+------------+------------+------------+


Classical evaluation:  -1.00 (white side)
NNUE evaluation:       1.54 (white side) [pure NNUE]
Final evaluation:      2.38 (white side) [with scaled NNUE, hybrid, ...]
```

**Comment by vondele** [comment]
I have one question on the signs of numbers are these consistent?

**Comment by snicolet** [comment]
There is opposite signs for black to play indeed, suggestions welcome

**Comment by snicolet** [comment]
output for Michel's position:

```
position fen 8/3k4/5P2/5K2/8/8/8/8 w - - 0 1
eval
info string NNUE evaluation using nn-50144f835024.nnue enabled

 Contributing terms for the classical eval:
+------------+-------------+-------------+-------------+
|    Term    |    White    |    Black    |    Total    |
|            |   MG    EG  |   MG    EG  |   MG    EG  |
+------------+-------------+-------------+-------------+
|   Material |  ----  ---- |  ----  ---- |  0.00  0.00 |
|  Imbalance |  ----  ---- |  ----  ---- |  0.00  0.00 |
|      Pawns |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|    Knights |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|    Bishops |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|      Rooks |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|     Queens |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|   Mobility |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|King safety |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|    Threats |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|     Passed |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|      Space |  0.00  0.00 |  0.00  0.00 |  0.00  0.00 |
|   Winnable |  ----  ---- |  ----  ---- |  0.00  0.00 |
+------------+-------------+-------------+-------------+
|      Total |  ----  ---- |  ----  ---- |  0.00  0.00 |
+------------+-------------+-------------+-------------+

 Pure NNUE derived piece values:
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |   k   |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   P   |       |       |
|       |       |       |       |       | +10.7 |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |   K   |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+
|       |       |       |       |       |       |       |       |
|       |       |       |       |       |       |       |       |
+-------+-------+-------+-------+-------+-------+-------+-------+

 Pure NNUE network contributions:
+------------+------------+------------+------------+
|   Bucket   |  Material  | Positional |   Total    |
|            |   (PSQT)   |  (Layers)  |            |
+------------+------------+------------+------------+
|  0         |  +  0.35   |  + 10.6o   |  + 10.9?   | <-- this bucket is used
|  1         |  +  0.75   |  +  5.50   |  +  6.25   |
|  2         |  +  1.15   |  +  2.83   |  +  3.98   |
|  3         |  +  1.45   |  +  2.50   |  +  3.96   |
|  4         |  +  1.56   |  +  1.91   |  +  3.47   |
|  5         |  +  0.86   |  +  0.97   |  +  1.83   |
|  6         |  +  0.13   |  +  1.42   |  +  1.55   |
|  7         |  +  0.64   |  +  1.59   |  +  2.24   |
+------------+------------+------------+------------+


Classical evaluation:  49.10 (white side)
NNUE evaluation:       10.99 (white side) [pure NNUE]
Final evaluation:      10.50 (white side) [with scaled NNUE, hybrid, ...]
```

**Comment by snicolet** [comment]
Which allows us to fix the last digit bug for values > 10.0, thanks :-)

**Comment by snicolet** [comment]
Maybe I can switch the signs in the "Pure NNUE network contributions" table when Black to play?

**Comment by vondele** [comment]
yes, that's probably a solution, or just mention in the caption/title the convention?

`+ 10.6o` that's an `o`  not a `0` 

```

Classical evaluation:  -1.00 (white side)
NNUE evaluation:       1.54 (white side) [pure NNUE]
Final evaluation:      2.38 (white side) [with scaled NNUE, hybrid, ...]
```

looks like the alignment is 1 off for the `.`

**Comment by Sopel97** [comment]
> There is opposite signs for black to play indeed, suggestions welcome

I think it should be for white to play, just as the evals at the end.
```
      if (pos.side_to_move() == BLACK) {
        t.psqt[bucket] *= -1;
        t.positional[bucket] *= -1;
      }
```
and I think it's better to replace the currently two places where eval is flipped for the board display with 
```
          v = pos.side_to_move() == WHITE ? v : -v;
```
after v is computed. I'll keep this branch as it was as I don't know how @snicolet is doing the changes, and I don't want to make it hard to merge.

> + 10.6o that's an o not a 0

yea, there's a bug
```
    else if (cp >= 1000) {
      buffer[1] = ' ';
      buffer[2] = '0' + cp / 1000; cp %= 1000;
      buffer[3] = '0' + cp / 100; cp %= 100;
      buffer[4] = '.';
      buffer[5] = '0' + cp / 10; // HERE cp %= 10; missing!
      buffer[6] = '0' + cp;
    }
```

**Comment by snicolet** [comment]
Merged via 2e745956c0478627c4822104fc48a079c2f15e77, thanks!

- use Stockfish coding style, fix some bugs
- added table headers
- renamed export_net() to save_eval() while there

@vdbergh 
(1) is harder to format because of the specialized eval indeed, I had a look but couldn't do it in 5 minutes
(2) and (3) I have tried to mitigate in the committed version, sort of :-)

**Pull request closed without merging**